Implement undelete and purge API for users

cganote opened the pull request:
Refactored the controller vs. manager for the deletion and purge of users. Purge now wipes out the username (regardless of gdrp setting) and takes out histories. Undelete should now work.

mvdbeek committed:
Add an API test for deleting users

Commit:
add api test

juleengraham committed:
Addressed lint issues

cganote committed:
Refactored the controller vs. manager for the deletion and purge of users. Purge now wipes out the username (regardless of gdrp setting) and takes out histories. Undelete should now work.

martenson commented:
@galaxybot test this

cganote committed:
Fixes, I hope!

hexylena commented:
> Purge now wipes out the username...and takes out histories.

:tada::heart::tada::heart::tada::heart:

martenson committed:
style fixes

cganote reviewed (commented):
Great, thank you!

martenson committed:
remove some extra debugs

martenson commented:
just now noticed https://github.com/galaxyproject/galaxy/pull/8299 :/

martenson committed:
minor refactoring

martenson committed:
fix tests

martenson committed:
bubble up exception without masking, adjust docstring

martenson commented:
I cherry picked the test commits from https://github.com/galaxyproject/galaxy/pull/8299 and fixed them. This PR now supersedes the other one.

martenson commented:
@galaxybot test this

hexylena commented:
Fixes https://github.com/galaxyproject/galaxy/issues/930

nsoranzo commented on lib/galaxy/managers/users.py:
```diff
@@ -109,10 +111,108 @@ def create(self, email=None, username=None, password=None, **kwargs):
+            log.debug("Cannot delete purged user, but setting delete flag just in case.")
         user.deleted = True
         self.session().add(user)
         self.session().flush()
```
I'd just replace lines 117-125 with:
```python
        super(UserManager, self).delete(user)
```

nsoranzo commented on lib/galaxy/managers/users.py:
```diff
@@ -109,10 +111,108 @@ def create(self, email=None, username=None, password=None, **kwargs):
+            log.debug('User \'%s\' has not been deleted, so it cannot be undeleted.' % user.email)
+        user.deleted = False
+        self.session().add(user)
+        self.session().flush()
```
I'd replace lines 133-137 with:
```python
        super(UserManager, self).undelete(user)
```

nsoranzo commented on lib/galaxy/managers/users.py:
```diff
@@ -109,10 +111,108 @@ def create(self, email=None, username=None, password=None, **kwargs):
+        if not self.app.config.allow_user_deletion:
+            raise exceptions.ConfigDoesNotAllowException('The configuration of this Galaxy instance does not allow admins to delete or purge users.')
+        if not user.deleted:
+            raise exceptions.MessageException('User \'%s\' has not been deleted, so they cannot be purged.' % user.email)
```
I'd remove lines 143-144, purging without deleting first is fine by me.

martenson replied:
I think it is a useful extra safety layer before a destructive function is called. @cganote was it meant like this?

cganote replied:
Yep I was hoping to have a little bit of a safety net

nsoranzo replied:
I don't think this check belongs to the manager.

martenson replied:
I think it is better here than in the API controller though. Where would you put it?

nsoranzo replied:
In the UI (`lib/galaxy/webapps/galaxy/controllers/admin.py` ?), I don't think it's the API business to ask users for confirmation or similar.

martenson replied:
This is not a confirmation, this is a built-in safety measure before a destructive operation happens. I think it is much aligned with how carefully Galaxy handles similar situations.

nsoranzo replied:
History purge doesn't have this check in either the manager or the API.

martenson replied:
Purge history is a user-centric operation that has an extra ui confirmation. Purge user is an admin operation, usually way more destructive, and exposed through API. I do not think the extra confirmation belongs just to the UI. I am not opposed to adding a UI check, but this PR does a good job at trying to minimize impact of potential admin errors.

nsoranzo commented on lib/galaxy/managers/users.py:
```diff
@@ -109,10 +111,108 @@ def create(self, email=None, username=None, password=None, **kwargs):
+                self.session().add(addr)
+
+        self.session().add(user)
+        self.session().flush()
```
I'd replace lines 166 and 213-214 with:
```python
        super(UserManager, self).purge(user)
```

bernt-matthias commented:
Any chance to cover this old one by me https://github.com/galaxyproject/galaxy/issues/3843

martenson commented:
@bernt-matthias I wouldn't tie it to this PR though

bernt-matthias commented:
> I wouldn't tie it to this PR though

sure

martenson committed:
use the superclass for basic operations

suggested by nsoranzo

martenson committed:
remove unused imports

martenson committed:
avoid calling superclass method for purging

Because multiple inheritance will result in calling this class 'delete' method
which is not the correct target. This also avoid a database operation
I believe, since the PurgableManagerMixin.purge also calls delete before purging.

nsoranzo commented on test/api/test_users.py:
```diff
@@ -74,6 +77,36 @@ def test_admin_update(self):
+        """Delete user and then purge them."""
+        user = self._setup_user(TEST_USER_EMAIL_PURGE)
+        response = self._delete("users/%s" % user["id"], admin=True)
+        self._assert_status_code_is(response, 200)
```
```suggestion
        self._assert_status_code_is_ok(response)
```

nsoranzo commented on test/api/test_users.py:
```diff
@@ -74,6 +77,36 @@ def test_admin_update(self):
+        self._assert_status_code_is(response, 200)
+        data = dict(purge="True")
+        response = self._delete("users/%s" % user["id"], data=data, admin=True)
+        self._assert_status_code_is(response, 200)
```
```suggestion
        self._assert_status_code_is_ok(response)
```

nsoranzo commented on lib/galaxy/managers/users.py:
```diff
@@ -109,7 +111,91 @@ def create(self, email=None, username=None, password=None, **kwargs):
+            raise exceptions.ItemDeletionException('Purged user cannot be undeleted.')
+        super(UserManager, self).undelete(user)
+
+    def purge(self, user):
```
```suggestion
    def purge(self, user, flush=True):
```

nsoranzo commented on lib/galaxy/managers/users.py:
```diff
@@ -109,7 +111,91 @@ def create(self, email=None, username=None, password=None, **kwargs):
         return user
 
     def delete(self, user):
```
```suggestion
    def delete(self, user, flush=True):
```

nsoranzo commented on lib/galaxy/managers/users.py:
```diff
@@ -109,7 +111,91 @@ def create(self, email=None, username=None, password=None, **kwargs):
+            raise exceptions.ConfigDoesNotAllowException('The configuration of this Galaxy instance does not allow admins to delete users.')
+        super(UserManager, self).delete(user)
+
+    def undelete(self, user):
```
```suggestion
    def undelete(self, user, flush=True):
```

nsoranzo commented on lib/galaxy/managers/users.py:
```diff
@@ -109,7 +111,91 @@ def create(self, email=None, username=None, password=None, **kwargs):
+        """Mark the given user deleted."""
+        if not self.app.config.allow_user_deletion:
+            raise exceptions.ConfigDoesNotAllowException('The configuration of this Galaxy instance does not allow admins to delete users.')
+        super(UserManager, self).delete(user)
```
```suggestion
        super(UserManager, self).delete(user, flush=flush)
```

nsoranzo commented on lib/galaxy/managers/users.py:
```diff
@@ -109,7 +111,91 @@ def create(self, email=None, username=None, password=None, **kwargs):
+            raise exceptions.ConfigDoesNotAllowException('The configuration of this Galaxy instance does not allow admins to undelete users.')
+        if user.purged:
+            raise exceptions.ItemDeletionException('Purged user cannot be undeleted.')
+        super(UserManager, self).undelete(user)
```
```suggestion
        super(UserManager, self).undelete(user, flush=flush)
```

nsoranzo commented on lib/galaxy/managers/users.py:
```diff
@@ -109,7 +111,91 @@ def create(self, email=None, username=None, password=None, **kwargs):
+                addr.phone = new_secure_hash(addr.phone + pseudorandom_value)
+                self.session().add(addr)
+        # Purge the user
+        user.purged = True
```
```suggestion
        super(UserManager, self).purge(user, flush=flush)
```

nsoranzo commented on lib/galaxy/managers/users.py:
```diff
@@ -109,7 +111,91 @@ def create(self, email=None, username=None, password=None, **kwargs):
+                self.session().add(addr)
+        # Purge the user
+        user.purged = True
         self.session().add(user)
```
```suggestion
```

nsoranzo commented on lib/galaxy/managers/users.py:
```diff
@@ -109,7 +111,91 @@ def create(self, email=None, username=None, password=None, **kwargs):
+        # Purge the user
+        user.purged = True
         self.session().add(user)
         self.session().flush()
```
```suggestion
```

martenson commented:
@nsoranzo I think I'm not a fan of going from api to manager then to mixin, then same manager again, then same mixin again for one method. It feels confusing so I'd like to avoid that.

nsoranzo commented:
> @nsoranzo I think I'm not a fan of going from api to manager then to mixin, then same manager again, then same mixin again for one method. It feels confusing so I'd like to avoid that.

I think these are good abstractions. For reference:
- the API takes care of the communication with the client, checks the inputs and permissions, then invokes the methods of the right manager
- the manager inherits and specialises some methods from superclasses and mixins, ensuring a consistent interface
- the `PurgableManagerMixin` mixin implements a generic `purge()` method, which just calls the `delete()` method and sets the `purged` attribute of the item
- for `UserManager`, we need to specialise both the `delete()` and `purge()` method, inherited from `PurgableManagerMixin`

martenson commented:
I don't think the path manager>mixin>manager>mixin>done  is better than manager>done but at this point it seems like a matter of opinion and I will defer to yours.

martenson committed:
incorporate @nsoranzo's abstraction suggestions

hexylena reviewed (approved):
Fine for me :)
